### ISSUE TITLE
Streaming Large Files

### ISSUE THREAD

**Issue by petermorrownavican** [tiangolo/fastapi#58]
I am working with [this example](https://fastapi.tiangolo.com/tutorial/request-files/#define-file-parameters) in the documentation that shows a basic example of how to work with files.

Is it possible to stream a large file with this? What is the correct pattern?

It's possible to [stream files with starlette](https://www.starlette.io/requests/#body), so how would integrating this functionality with fastapi work?

Thank you!

**Comment by matthdsm** [comment]
Seconded!
I'm also trying to get streaming responses to work, but I haven't had any success so far.
I'm trying to stream a (large) json response to a reactive front-end (vue), which should render the data as is becomes available.

Thanks for any pointers you might have.
Cheers
M

**Comment by tiangolo** [comment]
First, for the case of @petermorrownavican , I added a new feature to support `UploadFile` (from Starlette) in FastAPI, the new documentation is here: https://fastapi.tiangolo.com/tutorial/request-files/#file-parameters-with-uploadfile

That would probably be the best way to deal with large files.

It uses a standard "spooled" file (in memory up to some limit, then on disk).

And you can interact with it as a normal file in Python.

## Stream **request** content

If you have some specific use case that requires you to read the `bytes` as a stream of content, chunk by chunk (that also means that you don't need to have to whole content/file before starting to read it), you can use the same code as in the example to stream a file with Starlette.

You just have to [declare a parameter to take the Starlette `Request`](https://fastapi.tiangolo.com/tutorial/using-request-directly/), and then the same:

```Python
async for chunk in request.stream():
   ...
```

Adapting the example in Starlette, it would be something similar to:

```Python
from fastapi import FastAPI
from starlette.requests import Request

app = FastAPI()


@app.post("/files/")
def create_file(request: Request):
    body = b''
    async for chunk in request.stream():
        body += chunk
    response = Response(body, media_type='text/plain')
    return response
```

## Stream **response** content

In the case of @matthdsm , you can return a [Starlette `StreamingResponse`](https://www.starlette.io/responses/#streamingresponse).

It would be similar to:

```Python
from fastapi import FastAPI
from starlette.requests import Request
from starlette.responses import StreamingResponse

app = FastAPI()


@app.get("/files/")
def read_stream():
    return StreamingResponse(some_generator, media_type='application/json')
```

But now, have in mind that JSON is a format that needs the whole file to be ready before being parsed. Maybe you have a custom receiver that takes JSON lines separated by new line characters or something similar, but otherwise, if it's pure JSON, your frontend won't be able to parse the contents until you have the full JSON downloaded.

If you need to send different arbitrary JSON messages/documents, you can also think of [using `WebSocket`s](https://fastapi.tiangolo.com/tutorial/websockets/).

**Comment by matthdsm** [comment]
Hi @tiangolo 

Thanks for the comprehensive explanation! I got it to work using the example you provided. My front-end will have to be able to receive a json stream, since I'm outputting json objects. I've tried using complete json documents, but in my case, that just doesn't work at all. 
I'll look into websockets, thanks for the suggestion!

Cheers
M

**Comment by petermorrownavican** [comment]
Thank you for your quick, thorough and thoughtful responses @tiangolo! The new feature to support `UploadFile` works perfectly for us.

**Comment by tiangolo** [comment]
Awesome! Thanks for reporting back and closing the issue :+1: :taco: 

Just out of curiosity, are you using FastAPI at NAVICAN?

**Comment by petermorrownavican** [comment]
We are experimenting with it in dev to see if it meets our needs, but not in production (yet). We like it a lot so far though! 👍 🌯

**Comment by tiangolo** [comment]
That's great to hear! I would love to know how it goes.

**Comment by cmliotta** [comment]
@tiangolo I am returning a StreamingResponse as suggested above `StreamingResponse(image, media_type='image/png')` and this works great, but wondering if there is a way to show the response in the documentation rather than showing "null"

<img width="509" alt="Screen Shot 2019-06-13 at 11 02 54 AM" src="https://user-images.githubusercontent.com/8262123/59456322-ef58df80-8dca-11e9-9a5e-51a8e8b9d56d.png">

**Comment by tiangolo** [comment]
@cmliotta yep, here are the docs: https://fastapi.tiangolo.com/advanced/custom-response/#document-in-openapi-and-override-response

Also here: https://fastapi.tiangolo.com/advanced/additional-responses/

**Comment by JarroVGIT** [comment]
Chunks are always sized in bytes. Any translation from len(bytes) -> seconds should be manually implemented. Wouldn’t know how though, I’d guess it would depend on the file format.

**Comment by sciai-ai** [comment]
@JarroVGIT 

1. where can i specify the number of bytes parameter for the chunks? The file format is wav 16Khz, 16 bits/sample. I want to have 15 seconds worth of bytes in a chunk.
2.  I tried writing the chunks to a wav file but I get an error because the chunk is actually a formdata object
`b'--------------------------e4462b3dae0007d5\r\nContent-Disposition: form-data; name="file"; filename="audio.wav"\r\nContent-Type: audio/wav\r\n\r\nRIFF$p\x02\x00WAVEfmt \x10\x00\x00\x00\x01\x00\x01\x00\x80>\x00\x00\x00}\x00\x00\x02\x00\x10\x00data\x00p\x02\x00\x00\x00\x00\x00\xff\xff\x00\x00\xff\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff\x00\x00\xff\xff`

How can I access the bytes in the chunk object, based on the code below

> ```python
> from fastapi import FastAPI
> from starlette.requests import Request
> 
> app = FastAPI()
> 
> 
> @app.post("/files/")
> def create_file(request: Request):
>     body = b''
>     async for chunk in request.stream():
>         body += chunk
>     response = Response(body, media_type='text/plain')
>     return response
> ```

**Comment by Yukissss** [comment]
I also wonder if we can set an actual chunk size when iter through the stream. Like the code below, if I am reading a large file like 4GB here and want to write the chunk into server's file, it will trigger too many operations that writing chunks into file if chunk size is small by default.

```
@app.post("/files/")
def upload_file(request: Request):
    with open(file,'wb') as f:
        async for chunk in request.stream():  
            f.write(chunk)
```
